### ISSUE TITLE
System.IO.Compression: Add IsFile or IsDirectory property to ZipArchiveEntry

### ISSUE THREAD

**Issue by DanAtkinson** [dotnet/corefx#19889]
The current mechanism to determine whether an item in a zip is a directory is to test whether the `FullName` ends with `"/"` and also the `Name` property is blank.

To my knowledge, in order to extract a zip file and overwrite the contents - since `ExtractToDirectory` doesn't have that ability yet (#8146!) - the following is the smallest way of performing this:

    $zip = [System.IO.Compression.ZipFile]::OpenRead("MyCompressedFile.zip")
    foreach ($item in $zip.Entries) {
      if ($item.FullName.EndsWith("/") -and $item.name -eq "") {
        (New-Item $item.FullName -type directory -force) > $null
      } else {
        ([System.IO.Compression.ZipFileExtensions]::ExtractToFile($item, (Join-Path -Path $pathToExtractDirectory -ChildPath $item.FullName), $true)) > $null
      }
    }

As you can see on the third line, we need to determine whether the `ZipArchiveEntry` is a directory or file - `$item.FullName.EndsWith("/") -and $item.name -eq ""`.  If this condition can be changed to `$item.IsDirectory` or `$item.IsFile` then I believe that this would make it easier for developers to extract files in this way.

The function of the `IsDirectory` property is simply to provide a shorthand, user-friendly mechanism of succinctly determining the object type of the `ZipArchiveEntry`. It could also be an enum, such as `ZipArchiveEntryType`, but I believe that this might be overkill.

**Comment by karelz** [comment]
@ianhays don't we have something like that already?

@DanAtkinson if it doesn't exist, we will need formal API proposal - ideally formatted as example in [API process](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/api-review-process.md) - it is just much easier to read and reason about. Thanks!

**Comment by DanAtkinson** [comment]
@karelz I did check the [docs](https://msdn.microsoft.com/en-us/library/system.io.compression.ziparchiveentry(v=vs.110).aspx) first but I can't see anything that would aid being able to differentiate between file and folder.

**Current Properties of ZipArchiveEntry**
Archive, CompressedLength, FullName, LastWriteTime, Length, Name

**Current Methods of ZipArchiveEntry:**
Delete(), Equals(Object), Finalize(), GetHashCode(), GetType(), MemberwiseClone(), Open(), ToString()

If required, I can knock together a proposal.

Thanks, Dan

**Comment by ianhays** [comment]
> since ExtractToDirectory doesn't have that ability yet (#8146!)

We added that function in https://github.com/dotnet/corefx/pull/14806. Note that is why #8146 has been closed.

>don't we have something like that already?

No, we do not. In ZipFile we just check for a trailing slash. I don't think the `IsDirectory`/`IsFile` Properties would save devs enough time/code to justify the additional API, particularly since our method of distinguishing between the two is based on trailing character. If the Zip spec carried a bool indicating whether an entry was a file or directory and we exposed it then that would make sense, but I would prefer not to expose an API that conflates our slash assumption as absolute fact.
